drivers/volume/portworx/portworx.go

```diff
@@ -261,42 +300,202 @@ func (p *portworx) getSnapshotName(tags *map[string]string) string {
-			namespaceLabel:     (*tags)[snapshotter.CloudSnapshotCreatedForVolumeSnapshotNamespaceTag],
-		},
-	}
-	snapshotID, err := p.volDriver.Snapshot(volumeID, true, locator)
+	snapType, err := getSnapshotType(snap)
 	if err != nil {
-		return nil, nil, err
+		return nil, getErrorSnapshotConditions(err), err
+	}
+
+	switch snapType {
+	case crdv1.PortworxSnapshotTypeCloud:
+		logrus.Debugf("Cloud SnapshotCreate for pv: %+v \n tags: %v", pv, tags)
+		snapshotCredID = getCredIDFromSnapshot(snap)
+		request := &api.CloudBackupCreateRequest{
+			VolumeID:       volumeID,
+			CredentialUUID: snapshotCredID,
+		}
+		err := p.volDriver.CloudBackupCreate(request)
+		if err != nil {
+			return nil, getErrorSnapshotConditions(err), err
+		}
+
+		logrus.Infof("cloudsnap backup started successfully. waiting for completion...")
+		err = wait.ExponentialBackoff(cloudsnapBackoff, func() (bool, error) {
+			snapshotID, err = p.checkCloudSnapStatus(volumeID)
+			if err != nil {
+				logrus.Error(err)
+				return false, nil
+			}
+
+			return true, nil
+		})
+		if err != nil {
+			return nil, getErrorSnapshotConditions(err), err
+		}
+	case crdv1.PortworxSnapshotTypeLocal:
+		if isGroupSnap(snap) {
+			groupID := snap.Metadata.Annotations[pxSnapshotGroupIDKey]
+			groupLabels := parseGroupLabelsFromAnnotations(snap.Metadata.Annotations)
+
+			infoStr := fmt.Sprintf("creating group snapshot: [%s] %s", snap.Metadata.Namespace, snap.Metadata.Name)
+			if len(groupID) > 0 {
+				infoStr = fmt.Sprintf("%s. Group ID: %s", infoStr, groupID)
+			}
+
+			if len(groupLabels) > 0 {
+				infoStr = fmt.Sprintf("%s. Group labels: %v", infoStr, groupLabels)
+			}
+
+			logrus.Infof(infoStr)
+
+			resp, err := p.volDriver.SnapshotGroup(groupID, groupLabels)
+			if err != nil {
+				return nil, getErrorSnapshotConditions(err), err
+			}
+
+			if len(resp.Snapshots) == 0 {
+				err = fmt.Errorf("found 0 snapshots using given group selector/ID")
+				return nil, getErrorSnapshotConditions(err), err
+			}
+
+			snapIDs := make([]string, 0)
+			snapDataNames := make([]string, 0)
+			failedSnapVols := make([]string, 0)
+
+			// Loop through the response and check if all succeeded. Don't create any k8s objects if
+			// any of the snapshots failed
+			for volID, newSnapResp := range resp.Snapshots {
+				if newSnapResp.GetVolumeCreateResponse() == nil {
+					err = fmt.Errorf("Portworx API returned empty response on creation of group snapshot")
+					return nil, getErrorSnapshotConditions(err), err
+				}
+
+				if newSnapResp.GetVolumeCreateResponse().GetVolumeResponse() != nil &&
+					len(newSnapResp.GetVolumeCreateResponse().GetVolumeResponse().GetError()) != 0 {
+					logrus.Errorf("Failed to create snapshot for volume: %s due to: %v",
+						volID, newSnapResp.GetVolumeCreateResponse().GetVolumeResponse().GetError())
+					failedSnapVols = append(failedSnapVols, volID)
+					continue
+				}
+			}
+
+			if len(failedSnapVols) > 0 {
```

Comment:
We need to do either one of the following:
1. If any snap fails, delete all the snaps
2. If any snap fails, create objects for the ones that passed and report errors for the ones that failed (in either the volumesnapshot or volumesnapshotdata object).

We can't leave dangling snaps